src/v/cloud_storage/async_manifest_view.cc

```diff
@@ -0,0 +1,353 @@
+}
+
+size_t materialized_manifest_cache::size() const noexcept {
+    return _access_order.size() + _eviction_rollback.size();
+}
+
+size_t materialized_manifest_cache::size_bytes() const noexcept {
+    size_t res = 0;
+    for (const auto& m : _access_order) {
+        res += m._units.count();
+    }
+    for (const auto& m : _eviction_rollback) {
+        res += m._units.count();
+    }
+    return res;
+}
+
+void materialized_manifest_cache::put(
+  ss::semaphore_units<> s, spillover_manifest manifest) {
+    auto so = manifest.get_start_offset();
+    vassert(
+      so.has_value(), "Manifest can't be empty, ntp: {}", manifest.get_ntp());
+    if (!_eviction_rollback.empty()) {
+        auto it = lookup_eviction_rollback_list(so.value());
+        if (it != _eviction_rollback.end()) {
+            vlog(
+              _ctxlog.error,
+              "Manifest with base offset {} is being evicted from the "
+              "cache",
+              so);
+            throw std::runtime_error(fmt_with_ctx(
+              fmt::format,
+              "Manifest with start offset {} is being evicted from the "
+              "cache",
+              so));
+        }
+    }
+    auto item = ss::make_shared<materialized_manifest>(
+      std::move(manifest), std::move(s));
+    auto [it, ok] = _cache.insert(std::make_pair(so.value(), std::move(item)));
+    if (!ok) {
+        // This may indicate a race, log a warning
+        vlog(
+          _ctxlog.error, "Manifest with base offset {} is already present", so);
+        return;
+    }
+    _access_order.push_back(*it->second);
+}
+
+ss::shared_ptr<materialized_manifest>
+materialized_manifest_cache::get(model::offset base_offset) {
+    if (auto it = _cache.find(base_offset); it != _cache.end()) {
+        if (promote(it->second)) {
+            return it->second;
+        }
+    }
+    if (!_eviction_rollback.empty()) {
+        // Another fiber is waiting for the eviction of some elements.
+        // These elements could be stored in the '_eviction_rollback' list
+        // until there exist a copy of the shared pointer somewhere. We need
+        // to search through the list and return matching manifest if
+        // possible. Otherwise, the fiber may re-create the manifest and the
+        // other fiber may restore evicted manifest (if the wait on a
+        // semaphore will timeout) which will result in conflict.
+        auto it = lookup_eviction_rollback_list(base_offset);
+        if (it != _eviction_rollback.end()) {
+            return it->shared_from_this();
+        }
+    }
+    return nullptr;
+}
+
+bool materialized_manifest_cache::promote(model::offset base) {
+    if (auto it = _cache.find(base); it != _cache.end()) {
+        return promote(it->second);
+    }
+    return false;
+}
+
+bool materialized_manifest_cache::promote(
+  ss::shared_ptr<materialized_manifest>& manifest) {
+    auto node = manifest->_hook.this_ptr();
+    for (auto& other : _access_order) {
```

Comment:
This is O(N), right?  This seems really expensive to do on each get()

I guess this logic is a consequence of handling the case where a materialized_manifest is in _eviction_rollback but not _access order -- otherwise we could just unlink+push_back unconditionally.

It seems like it would be less onerous to use a `vector` inside `prepare` for the list of things to roll back, to keep the main get/promote path fast.  The list of manifests to roll back should always be small because we're only evicting enough to allow one more manifest in.